src/Trie.mo

```diff
@@ -129,44 +131,23 @@ module {
     }
   };
 
-  /// Checks the invariants of the trie structure, including the placement of keys at trie paths
-  public func isValid<K, V>(t : Trie<K, V>, enforceNormal : Bool) : Bool {
-    func rec(t : Trie<K, V>, bitpos : ?Hash.Hash, bits : Hash.Hash, mask : Hash.Hash) : Bool {
+  public func assertIsValid<K, V>(t : Trie<K, V>) {
+    func rec(t : Trie<K, V>, bitpos : ?Hash.Hash, bits : Hash.Hash, mask : Hash.Hash) {
       switch t {
-        case (#empty) {
-          switch bitpos {
-            case null { true };
-            case (?_) { not enforceNormal };
-          }
-        };
+        case (#empty) { };
         case (#leaf(l)) {
           let len = List.size(l.keyvals);
-          ((len <= MAX_LEAF_SIZE) or (not enforceNormal))
-          and
-          len == l.size
-          and
-          ( List.all(
+          assert
+            len <= MAX_LEAF_SIZE + 1;
```

Comment:
Turns out that the `Trie` actually creates chains at the `#leaf` positions that are up to `MAX_LEAF_SIZE + 1`, probably because of an off-by-one error in the `put` logic.

Reply:
That issue is separate from the one for this PR, which concerns the behavior of `filter` and `mapFilter`.

This new invariant test passes for those operations now.

This new invariant test is also updated to reflect the insight of @nomeata in the discussion of #427 regarding the invariants of the `Trie`.  (The prior invariant test reflected some wrong assumptions, in part.)

Reply:
That's great, but if you suspect there's a bug in `put`, why not try to fix it now rather than changing the invariant.

I've tried to look myself but can't find it.